homeassistant/components/crownstone/config_flow.py

```diff
@@ -0,0 +1,113 @@
+        self.spheres = []
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=vol.Schema(
+                    {vol.Required(CONF_EMAIL): str, vol.Required(CONF_PASSWORD): str}
+                ),
+            )
+
+        self.cloud = CrownstoneCloud(
+            email=user_input[CONF_EMAIL],
+            password=user_input[CONF_PASSWORD],
+            websession=aiohttp_client.async_get_clientsession(self.hass),
+        )
+
+        # handle login errors on setup form
+        try:
+            await self.cloud.async_login()
+            # save email and password for later use
+            self.login_info = user_input
+            # start next flow
+            return await self.async_step_sphere()
+        except CrownstoneAuthenticationError as auth_error:
+            if auth_error.type == "LOGIN_FAILED":
+                errors["base"] = "invalid_auth"
+            if auth_error.type == "LOGIN_FAILED_EMAIL_NOT_VERIFIED":
+                errors["base"] = "account_not_verified"
+        except CrownstoneUnknownError:
+            errors["base"] = "unknown"
+
+        # show form again, with the error
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_EMAIL): str, vol.Required(CONF_PASSWORD): str}
+            ),
+            errors=errors,
+        )
+
+    async def async_step_sphere(self, user_input=None):
+        """Handle the step for selecting a sphere."""
+        # get the spheres for the user
+        await self.cloud.spheres.async_update_sphere_data()
+
+        # only 1 sphere configured, don't show form and set this as sphere
+        if len(self.cloud.spheres.spheres) == 1:
+            user_input = {
+                CONF_SPHERE: next(iter(self.cloud.spheres.spheres.values())).name
+            }
+
+        # show form with drop down menu
+        if user_input is None:
+            # generate sphere list
+            for sphere in self.cloud.spheres:
+                self.spheres.append(sphere.name)
+
+            return self.async_show_form(
+                step_id="sphere",
+                data_schema=vol.Schema({CONF_SPHERE: vol.In(self.spheres)}),
+            )
+
+        # set the unique id
+        await self.async_set_unique_id(user_input[CONF_SPHERE])
+        # make sure this sphere is only set up once
+        self._abort_if_unique_id_configured()
+
+        # cleanup RequestHandler
+        self.cloud.reset()
+
+        # return data to main
+        return self.async_create_entry(
+            title=self.unique_id,
+            data={
+                CONF_ID: self.unique_id,
+                CONF_EMAIL: self.login_info[CONF_EMAIL],
+                CONF_PASSWORD: self.login_info[CONF_PASSWORD],
```

Comment:
Since you're involved with Crownstone, can you ask them to update their authentication to use something like OAuth2? It's highly prefered over storing email/password. (not for this PR)